packages/rtop/test/ops_lib/TOpPairWiseMax_UnitTests.cpp

```diff
@@ -0,0 +1,118 @@
+//       Operations
+//                Copyright (2006) Sandia Corporation
+// 
+// Under terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+// license for use of this work by or on behalf of the U.S. Government.
+// 
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Roscoe A. Bartlett (rabartl@sandia.gov) 
+// 
+// ***********************************************************************
+// @HEADER
+*/
+
+
+#include "RTOpPack_TOpPairWiseMax.hpp"
+
+#include "supportUnitTestsHelpers.hpp"
+
+
+namespace {
+
+
+template<class Scalar>
+void basicTest(const int stride, FancyOStream &out, bool &success)
+{
+
+  using Teuchos::as;
+  typedef ScalarTraits<Scalar> ST;
+  typedef typename ST::magnitudeType ScalarMag;
+
+  ConstSubVectorView<Scalar> v0 =
+    newStridedRandomSubVectorView<Scalar>(n, stride);
+
+  ConstSubVectorView<Scalar> v1 =
+    newStridedRandomSubVectorView<Scalar>(n, stride);
+
+  SubVectorView<Scalar> orig_z0 =
+    newStridedRandomSubVectorView<Scalar>(n, stride);
+  
+  SubVectorView<Scalar> z0 =
+    newStridedSubVectorView<Scalar>(n, stride, ST::nan());
+  RTOpPack::assign_entries<Scalar>( Teuchos::outArg(z0), orig_z0 );
+
+  const Scalar alpha = 2.0*ST::one();
+
+  RTOpPack::TOpPairWiseMax<Scalar> op(alpha);
+  op.apply_op( tuple(v0,v1), tuple(z0)(), null );
+
+  SubVectorView<Scalar> expected_z0
+    = newStridedSubVectorView<Scalar>(n, stride, ST::nan());
+  for (int k = 0; k < v0.subDim(); ++k)
+    expected_z0[k] =  v0[k] + v1[k] + std::abs(v0[k] - v1[k]);
```

Comment:
I am a bit confused by this.  Should this not just be:

```
expected_z0[k] =  alpha * std::max(v0[k] - v1[k]);
```

?

Reply:
Should not the computation of `expected_z0` not involve `alpha`?

Okay, I see, this is using the factor that `alpha==2` and does not work for any other `alpha`.

Why have a a test like this?  If someone changes `alpha` the test will break.  Why not just write a brain-dead but super strong test like: 

```
n = 2;
alpha = 2.0;

v0[0] = 0.5;      v1[0] = 0;   expected_z0[0] = 1.0;
v0[1] = -0.25;   v1[1] = 0.3;  expected_z0[1] = 0.6;
v0[2] = -0.7;   v1[2] = -4.0;  expected_z0[2] = -1.4;
v0[3] = -4.0;   v1[3] = -0.8;  expected_z0[3] = -1.6;

op.apply_op( tuple(v0,v1), tuple(z0)(), null );

compare(expected_z0, z0);
```

That test is super strong, super obvious and will catch stupid copy and paste errors.  But it also might be better to use `alpha != 2.0` since it will catch a bad implementation.

Reply:
@bartlettroscoe I was trying to follow other TOp unittests to include the stride and nonstride tests. If you don't mind, perhaps could you show me an example demonstrating how I can test stride/nonstride with the TOpPairWiseMaxUpdate.